From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Andrew den Exter <andrew.den.exter@jolla.com>
Date: Thu, 11 Nov 2021 06:51:59 +0000
Subject: [PATCH 60/67] Disallow page zooming if the meta viewport scale is
 fixed. JB#56149 OMP#JOLLA-49

Zooming is always allowed if desktop mode is active. JB#56767
OMP#JOLLA-609
---
 layout/base/nsLayoutUtils.cpp | 25 ++++++++++++++++++++++---
 1 file changed, 22 insertions(+), 3 deletions(-)

diff --git a/layout/base/nsLayoutUtils.cpp b/layout/base/nsLayoutUtils.cpp
index 5963a74ab2c3..6f34f50de087 100644
--- a/layout/base/nsLayoutUtils.cpp
+++ b/layout/base/nsLayoutUtils.cpp
@@ -35,6 +35,7 @@
 #include "mozilla/DisplayPortUtils.h"
 #include "mozilla/glean/GleanMetrics.h"
 #include "mozilla/dom/AnonymousContent.h"
+#include "mozilla/dom/BrowsingContext.h"
 #include "mozilla/dom/BrowserChild.h"
 #include "mozilla/dom/CanvasUtils.h"
 #include "mozilla/dom/Document.h"
@@ -708,6 +709,23 @@ bool nsLayoutUtils::AsyncPanZoomEnabled(const nsIFrame* aFrame) {
   return widget->AsyncPanZoomEnabled();
 }
 
+static bool DocumentHasFixedUnityZoom(const mozilla::dom::Document* aDocument)
+{
+  if (!aDocument) {
+    return false;
+  }
+  ViewportMetaData metaData = aDocument->GetViewportMetaData();
+  return (!metaData.mMinimumScale.IsEmpty() && metaData.mMinimumScale == metaData.mMaximumScale) ||
+          metaData.mUserScalable.EqualsLiteral("0") ||
+          metaData.mUserScalable.EqualsLiteral("no") ||
+          metaData.mUserScalable.EqualsLiteral("false");
+}
+
+static bool IsDesktopView(const mozilla::dom::Document* aDocument) {
+  BrowsingContext* bc = aDocument ? aDocument->GetBrowsingContext() : nullptr;
+  return bc && bc->Top()->ForceDesktopViewport();
+}
+
 bool nsLayoutUtils::AllowZoomingForDocument(
     const mozilla::dom::Document* aDocument) {
   if (aDocument->GetPresShell() &&
@@ -717,10 +735,11 @@ bool nsLayoutUtils::AllowZoomingForDocument(
   // True if we allow zooming for all documents on this platform, or if we are
   // in RDM and handling meta viewports, which force zoom under some
   // circumstances.
+
   BrowsingContext* bc = aDocument ? aDocument->GetBrowsingContext() : nullptr;
-  return StaticPrefs::apz_allow_zooming() ||
-         (bc && bc->InRDMPane() &&
-          nsLayoutUtils::ShouldHandleMetaViewport(aDocument));
+  return (StaticPrefs::apz_allow_zooming() && !DocumentHasFixedUnityZoom(aDocument)) ||
+         (bc && bc->InRDMPane() && nsLayoutUtils::ShouldHandleMetaViewport(aDocument)) ||
+         IsDesktopView(aDocument);
 }
 
 static bool HasVisibleAnonymousContents(Document* aDoc) {
